src/EventStore.Core/Messaging/TcsEnvelope.cs

```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using System.Threading.Tasks;
+
+namespace EventStore.Core.Messaging {
+	public class TcsEnvelope<TResult> : IEnvelope where TResult : class {
+		private readonly TaskCompletionSource<TResult> _tcs;
+		public Task<TResult> Task => _tcs.Task;
+
+		public TcsEnvelope() {
+			_tcs = new TaskCompletionSource<TResult>(TaskCreationOptions.RunContinuationsAsynchronously);
+		}
+
+		public void ReplyWith<T>(T message) where T : Message {
+			if (message is not TResult result)
+				throw new ArgumentException("Unexpected message type.", nameof(message));
+
+			_tcs.SetResult(result);
```

Comment:
```suggestion
			_tcs.TrySetResult(result);
```

perhaps better to use TrySetResult, after all it isn't the fault of the thing replying if something else has also subscribed to the message and already replied

Reply:
fixed :+1: